File: internal/configbackup/capture.go
package configbackup

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"strings"

	"github.com/google/uuid"
)

// CaptureConfig copies every allowlisted user-authored file under root into
// <appDataDir>/config-backups/<uuid>/ (0o700 dirs, mode-preserving copy),
// SHA-256s each, and writes manifest.json (0o600). Symlinked skills are recorded
// as SkillLink refs (target string only, NO content — an out-of-root repo is a
// documented non-goal). When includeHooksDisabled is set (the pre-import
// auto-snapshot), it ALSO snapshots <appDataDir>/hooks-disabled.json so a
// one-click undo fully reverts the disabled groups an import appended.
func CaptureConfig(root, appDataDir, label string, includeHooksDisabled bool) (Manifest, error) {
	canonRoot, err := filepath.EvalSymlinks(root)
	if err != nil {
		return Manifest{}, fmt.Errorf("configbackup: resolve root %q: %w", root, err)
	}

	id := uuid.NewString()
	backupDir := filepath.Join(configBackupsDir(appDataDir), id)
	if err := os.MkdirAll(backupDir, 0o700); err != nil {
		return Manifest{}, fmt.Errorf("configbackup: create backup dir: %w", err)
	}

	rels, skillLinks := collectConfigFiles(canonRoot)

	manifest := Manifest{
		ID:              id,
		Label:           label,
		CreatedMs:       nowMS(),
		SecretsIncluded: false,
		Files:           []FileEntry{},
		SkillLinks:      skillLinks,
	}
	for _, rel := range rels {
		entry, err := copyCaptureFile(canonRoot, backupDir, rel)
		if err != nil {
			return Manifest{}, err
		}
		manifest.Files = append(manifest.Files, entry)
	}

	if includeHooksDisabled {
		if err := captureHooksDisabled(appDataDir, backupDir); err != nil {
			return Manifest{}, err
		}
	}

	if err := writeManifest(backupDir, manifest); err != nil {
		return Manifest{}, err
	}
	return manifest, nil
}

// collectConfigFiles enumerates every allowlisted file's root-relative path
// under canonRoot plus the symlinked-skill link refs. Every candidate is
// filtered through matchConfigAllowlist so capture can never diverge from
// restore/import.
func collectConfigFiles(canonRoot string) (rels []string, skillLinks []SkillLink) {
	for _, name := range []string{"settings.json", "CLAUDE.md", "RTK.md"} {
		if isRegularFile(filepath.Join(canonRoot, name)) {
			rels = append(rels, name)
		}
	}
	for _, dir := range []string{"rules", "commands", "tools"} {
		rels = append(rels, walkAllowlistedFiles(canonRoot, filepath.Join(canonRoot, dir))...)
	}
	rels = append(rels, listMarkdownFilesRel(canonRoot, filepath.Join(canonRoot, "agents"))...)
	rels = append(rels, collectMemoryFiles(canonRoot, "projects", "memory")...)
	rels = append(rels, collectMemoryFiles(canonRoot, "agent-memory", "")...)

	skillRels, links := collectSkillFiles(canonRoot)
	rels = append(rels, skillRels...)
	return rels, links
}

// collectMemoryFiles lists *.md under <parent>/<name>/<sub> for every child of
// <parent> (sub "" means directly under <name>). Covers projects/*/memory/*.md
// and agent-memory/*/*.md.
func collectMemoryFiles(canonRoot, parent, sub string) []string {
	base := filepath.Join(canonRoot, parent)
	entries, err := os.ReadDir(base)
	if err != nil {
		return nil
	}
	var out []string
	for _, e := range entries {
		if !e.IsDir() || strings.HasPrefix(e.Name(), ".") {
			continue
		}
		dir := filepath.Join(base, e.Name())
		if sub != "" {
			dir = filepath.Join(dir, sub)
		}
		out = append(out, listMarkdownFilesRel(canonRoot, dir)...)
	}
	return out
}

// collectSkillFiles captures real skill dirs (SKILL.md + references/**) and
// records symlinked skills as link refs, NEVER following a link (its target may
// live outside root).
func collectSkillFiles(canonRoot string) (rels []string, links []SkillLink) {
	skillsBase := filepath.Join(canonRoot, "skills")
	entries, err := os.ReadDir(skillsBase)
	if err != nil {
		return nil, nil
	}
	for _, e := range entries {
		name := e.Name()
		if len(name) == 0 || name[0] == '.' {
			continue
		}
		linkPath := filepath.Join(skillsBase, name)
		lst, err := os.Lstat(linkPath)
		if err != nil {
			continue
		}
		if lst.Mode()&os.ModeSymlink != 0 {
			if target, err := os.Readlink(linkPath); err == nil {
				links = append(links, SkillLink{Name: name, Target: target})
			}
			continue
		}
		if !lst.IsDir() {
			continue
		}
		if skillMd := filepath.Join(skillsBase, name, "SKILL.md"); isRegularFile(skillMd) {
			if rel, err := filepath.Rel(canonRoot, skillMd); err == nil && matchConfigAllowlist(rel) {
				rels = append(rels, rel)
			}
		}
		rels = append(rels, walkAllowlistedFiles(canonRoot, filepath.Join(skillsBase, name, "references"))...)
	}
	return rels, links
}

// walkAllowlistedFiles recursively lists every regular allowlisted file under
// absDir (skipping symlinks and .bak/.tmp byproducts), returning root-relative
// paths. A missing dir yields nothing.
func walkAllowlistedFiles(canonRoot, absDir string) []string {
	var out []string
	_ = filepath.WalkDir(absDir, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return nil // missing/unreadable subtree — skip
		}
		if d.IsDir() || d.Type()&fs.ModeSymlink != 0 || isBakTmp(d.Name()) {
			return nil
		}
		rel, relErr := filepath.Rel(canonRoot, path)
		if relErr != nil || !matchConfigAllowlist(rel) {
			return nil
		}
		out = append(out, rel)
		return nil
	})
	return out
}

// listMarkdownFilesRel lists non-symlink *.md files directly in absDir
// (non-recursive), returning allowlisted root-relative paths.
func listMarkdownFilesRel(canonRoot, absDir string) []string {
	entries, err := os.ReadDir(absDir)
	if err != nil {
		return nil
	}
	var out []string
	for _, e := range entries {
		if e.IsDir() || e.Type()&fs.ModeSymlink != 0 || !strings.HasSuffix(e.Name(), ".md") {
			continue
		}
		rel, relErr := filepath.Rel(canonRoot, filepath.Join(absDir, e.Name()))
		if relErr != nil || !matchConfigAllowlist(rel) {
			continue
		}
		out = append(out, rel)
	}
	return out
}

// copyCaptureFile copies canonRoot/rel into backupDir/rel (0o700 parents,
// mode-preserving) and returns its FileEntry with the SHA-256 of the bytes.
func copyCaptureFile(canonRoot, backupDir, rel string) (FileEntry, error) {
	src := filepath.Join(canonRoot, rel)
	info, err := os.Lstat(src)
	if err != nil {
		return FileEntry{}, fmt.Errorf("configbackup: stat %q: %w", rel, err)
	}
	data, err := os.ReadFile(src)
	if err != nil {
		return FileEntry{}, fmt.Errorf("configbackup: read %q: %w", rel, err)
	}
	dest := filepath.Join(backupDir, rel)
	if err := os.MkdirAll(filepath.Dir(dest), 0o700); err != nil {
		return FileEntry{}, fmt.Errorf("configbackup: mkdir for %q: %w", rel, err)
	}
	mode := info.Mode().Perm()
	if err := os.WriteFile(dest, data, mode); err != nil {
		return FileEntry{}, fmt.Errorf("configbackup: write %q: %w", rel, err)
	}
	if err := os.Chmod(dest, mode); err != nil { // WriteFile mode is umask-masked; force-preserve.
		return FileEntry{}, fmt.Errorf("configbackup: chmod %q: %w", rel, err)
	}
	sum := sha256.Sum256(data)
	return FileEntry{RelPath: rel, Size: int64(len(data)), SHA256: hex.EncodeToString(sum[:])}, nil
}

// captureHooksDisabled snapshots <appDataDir>/hooks-disabled.json into the
// backup under the reserved name, if it exists. A missing file is a no-op.
func captureHooksDisabled(appDataDir, backupDir string) error {
	data, err := os.ReadFile(filepath.Join(appDataDir, "hooks-disabled.json"))
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return fmt.Errorf("configbackup: read hooks-disabled.json: %w", err)
	}
	if err := os.WriteFile(filepath.Join(backupDir, hooksDisabledSnapshotName), data, 0o600); err != nil {
		return fmt.Errorf("configbackup: snapshot hooks-disabled.json: %w", err)
	}
	return nil
}

// isRegularFile reports whether path is a non-symlink regular file.
func isRegularFile(path string) bool {
	info, err := os.Lstat(path)
	return err == nil && info.Mode().IsRegular()
}

// isBakTmp reports whether name is a write-primitive .bak/.tmp byproduct.
func isBakTmp(name string) bool {
	return strings.HasSuffix(name, ".bak") || strings.HasSuffix(name, ".tmp")
}

File: src-tauri/src/configbackup/capture.rs
//! Ports `internal/configbackup/capture.go` — copies every allowlisted
//! user-authored file under `root` into `<app_data_dir>/config-backups/<uuid>/`
//! (0o700 dirs, mode-preserving copy), SHA-256s each, and writes manifest.json
//! (0o600). Symlinked skills are recorded as `SkillLink` refs (target string
//! only, NO content — an out-of-root repo is a documented non-goal). Symlinks in
//! walks are never followed; `.bak`/`.tmp` byproducts and dot-dirs are skipped.

use std::fs;
use std::io;
use std::path::Path;

use sha2::{Digest, Sha256};
use uuid::Uuid;

use super::store::{mkdir_all_mode, set_mode, write_file_mode, write_manifest};
use super::types::{
    config_backups_dir, match_config_allowlist, now_ms, FileEntry, Manifest, SkillLink,
    HOOKS_DISABLED_SNAPSHOT_NAME,
};

/// Copies every allowlisted user-authored file under `root` into the backup
/// store, SHA-256s each, and writes the manifest. When `include_hooks_disabled`
/// is set (the pre-import auto-snapshot), it ALSO snapshots
/// `<app_data_dir>/hooks-disabled.json`. Mirrors `CaptureConfig`.
pub fn capture_config(
    root: &Path,
    app_data_dir: &Path,
    label: &str,
    include_hooks_disabled: bool,
) -> Result<Manifest, String> {
    let canon_root = fs::canonicalize(root)
        .map_err(|e| format!("configbackup: resolve root {}: {e}", root.display()))?;

    let id = Uuid::new_v4().to_string();
    let backup_dir = config_backups_dir(app_data_dir).join(&id);
    mkdir_all_mode(&backup_dir, 0o700).map_err(|e| format!("configbackup: create backup dir: {e}"))?;

    let (rels, skill_links) = collect_config_files(&canon_root);

    let mut manifest = Manifest {
        id: id.clone(),
        label: label.to_string(),
        created_ms: now_ms(),
        secrets_included: false,
        files: Vec::new(),
        skill_links,
    };
    for rel in &rels {
        let entry = copy_capture_file(&canon_root, &backup_dir, rel)?;
        manifest.files.push(entry);
    }

    if include_hooks_disabled {
        capture_hooks_disabled(app_data_dir, &backup_dir)?;
    }

    write_manifest(&backup_dir, &manifest)?;
    Ok(manifest)
}

/// Enumerates every allowlisted file's root-relative path under `canon_root`
/// plus the symlinked-skill link refs. Mirrors `collectConfigFiles`.
fn collect_config_files(canon_root: &Path) -> (Vec<String>, Vec<SkillLink>) {
    let mut rels: Vec<String> = Vec::new();
    let mut skill_links: Vec<SkillLink> = Vec::new();

    for name in ["settings.json", "CLAUDE.md", "RTK.md"] {
        if is_regular_file(&canon_root.join(name)) {
            rels.push(name.to_string());
        }
    }
    for dir in ["rules", "commands", "tools"] {
        walk_allowlisted_files(canon_root, &canon_root.join(dir), &mut rels);
    }
    list_markdown_files_rel(canon_root, &canon_root.join("agents"), &mut rels);
    collect_memory_files(canon_root, "projects", "memory", &mut rels);
    collect_memory_files(canon_root, "agent-memory", "", &mut rels);
    collect_skill_files(canon_root, &mut rels, &mut skill_links);

    (rels, skill_links)
}

/// Lists *.md under `<parent>/<name>/<sub>` for every child dir of `<parent>`
/// (`sub` "" means directly under `<name>`). Mirrors `collectMemoryFiles`.
fn collect_memory_files(canon_root: &Path, parent: &str, sub: &str, out: &mut Vec<String>) {
    let base = canon_root.join(parent);
    for entry in sorted_dir(&base) {
        let Ok(ft) = entry.file_type() else { continue };
        let name = entry.file_name();
        let name = name.to_string_lossy();
        if !ft.is_dir() || name.starts_with('.') {
            continue;
        }
        let mut dir = base.join(&*name);
        if !sub.is_empty() {
            dir = dir.join(sub);
        }
        list_markdown_files_rel(canon_root, &dir, out);
    }
}

/// Captures real skill dirs (SKILL.md + references/**) and records symlinked
/// skills as link refs, NEVER following a link. Mirrors `collectSkillFiles`.
fn collect_skill_files(canon_root: &Path, rels: &mut Vec<String>, links: &mut Vec<SkillLink>) {
    let skills_base = canon_root.join("skills");
    for entry in sorted_dir(&skills_base) {
        let name = entry.file_name();
        let name = name.to_string_lossy();
        if name.is_empty() || name.starts_with('.') {
            continue;
        }
        let link_path = skills_base.join(&*name);
        let Ok(lst) = fs::symlink_metadata(&link_path) else {
            continue;
        };
        if lst.file_type().is_symlink() {
            if let Ok(target) = fs::read_link(&link_path) {
                links.push(SkillLink {
                    name: name.into_owned(),
                    target: target.to_string_lossy().into_owned(),
                });
            }
            continue;
        }
        if !lst.is_dir() {
            continue;
        }
        let skill_md = skills_base.join(&*name).join("SKILL.md");
        if is_regular_file(&skill_md) {
            if let Some(rel) = rel_path(canon_root, &skill_md) {
                if match_config_allowlist(&rel) {
                    rels.push(rel);
                }
            }
        }
        walk_allowlisted_files(canon_root, &skills_base.join(&*name).join("references"), rels);
    }
}

/// Recursively lists every regular allowlisted file under `abs_dir` (skipping
/// symlinks and .bak/.tmp byproducts), returning root-relative paths. A missing
/// dir yields nothing. Mirrors `walkAllowlistedFiles`.
fn walk_allowlisted_files(canon_root: &Path, abs_dir: &Path, out: &mut Vec<String>) {
    for entry in sorted_dir(abs_dir) {
        let Ok(ft) = entry.file_type() else { continue };
        if ft.is_symlink() {
            continue; // never follow
        }
        let path = entry.path();
        if ft.is_dir() {
            walk_allowlisted_files(canon_root, &path, out);
            continue;
        }
        let name = entry.file_name();
        if is_bak_tmp(&name.to_string_lossy()) {
            continue;
        }
        if let Some(rel) = rel_path(canon_root, &path) {
            if match_config_allowlist(&rel) {
                out.push(rel);
            }
        }
    }
}

/// Lists non-symlink *.md files directly in `abs_dir` (non-recursive), returning
/// allowlisted root-relative paths. Mirrors `listMarkdownFilesRel`.
fn list_markdown_files_rel(canon_root: &Path, abs_dir: &Path, out: &mut Vec<String>) {
    for entry in sorted_dir(abs_dir) {
        let Ok(ft) = entry.file_type() else { continue };
        let name = entry.file_name();
        let name = name.to_string_lossy();
        if ft.is_dir() || ft.is_symlink() || !name.ends_with(".md") {
            continue;
        }
        if let Some(rel) = rel_path(canon_root, &abs_dir.join(&*name)) {
            if match_config_allowlist(&rel) {
                out.push(rel);
            }
        }
    }
}

/// Copies `canon_root/rel` into `backup_dir/rel` (0o700 parents, mode-preserving)
/// and returns its FileEntry with the SHA-256 of the bytes. Mirrors
/// `copyCaptureFile`.
fn copy_capture_file(canon_root: &Path, backup_dir: &Path, rel: &str) -> Result<FileEntry, String> {
    let src = canon_root.join(rel);
    let info = fs::symlink_metadata(&src).map_err(|e| format!("configbackup: stat {rel:?}: {e}"))?;
    let data = fs::read(&src).map_err(|e| format!("configbackup: read {rel:?}: {e}"))?;

    let dest = backup_dir.join(rel);
    if let Some(parent) = dest.parent() {
        mkdir_all_mode(parent, 0o700).map_err(|e| format!("configbackup: mkdir for {rel:?}: {e}"))?;
    }
    let mode = mode_perm(&info);
    write_file_mode(&dest, &data, mode).map_err(|e| format!("configbackup: write {rel:?}: {e}"))?;
    // WriteFile mode is umask-masked; force-preserve like Go's os.Chmod.
    set_mode(&dest, mode).map_err(|e| format!("configbackup: chmod {rel:?}: {e}"))?;

    let mut hasher = Sha256::new();
    hasher.update(&data);
    Ok(FileEntry {
        rel_path: rel.to_string(),
        size: data.len() as i64,
        sha256: hex_encode(&hasher.finalize()),
    })
}

/// Snapshots `<app_data_dir>/hooks-disabled.json` into the backup under the
/// reserved name, if it exists. A missing file is a no-op. Mirrors
/// `captureHooksDisabled`.
fn capture_hooks_disabled(app_data_dir: &Path, backup_dir: &Path) -> Result<(), String> {
    let data = match fs::read(app_data_dir.join("hooks-disabled.json")) {
        Ok(d) => d,
        Err(e) if e.kind() == io::ErrorKind::NotFound => return Ok(()),
        Err(e) => return Err(format!("configbackup: read hooks-disabled.json: {e}")),
    };
    write_file_mode(&backup_dir.join(HOOKS_DISABLED_SNAPSHOT_NAME), &data, 0o600)
        .map_err(|e| format!("configbackup: snapshot hooks-disabled.json: {e}"))
}

/// Reports whether `path` is a non-symlink regular file. Mirrors `isRegularFile`.
fn is_regular_file(path: &Path) -> bool {
    match fs::symlink_metadata(path) {
        Ok(meta) => meta.file_type().is_file(),
        Err(_) => false,
    }
}

/// Reports whether `name` is a write-primitive .bak/.tmp byproduct. Mirrors
/// `isBakTmp`.
fn is_bak_tmp(name: &str) -> bool {
    name.ends_with(".bak") || name.ends_with(".tmp")
}

/// `filepath.Rel(canon_root, path)` for a `path` always contained under
/// `canon_root`, yielding a `/`-separated root-relative string.
fn rel_path(canon_root: &Path, path: &Path) -> Option<String> {
    path.strip_prefix(canon_root)
        .ok()
        .map(|p| p.to_string_lossy().into_owned())
}

/// Reads a directory into a filename-sorted vec of entries (mirrors Go's
/// `os.ReadDir` order); a missing/unreadable dir yields an empty vec.
fn sorted_dir(dir: &Path) -> Vec<fs::DirEntry> {
    let Ok(read_dir) = fs::read_dir(dir) else {
        return Vec::new();
    };
    let mut entries: Vec<_> = read_dir.filter_map(|e| e.ok()).collect();
    entries.sort_by_key(|e| e.file_name());
    entries
}

#[cfg(unix)]
fn mode_perm(meta: &fs::Metadata) -> u32 {
    use std::os::unix::fs::PermissionsExt;
    meta.permissions().mode() & 0o777
}

#[cfg(not(unix))]
fn mode_perm(_meta: &fs::Metadata) -> u32 {
    0o644
}

fn hex_encode(bytes: &[u8]) -> String {
    bytes.iter().map(|b| format!("{b:02x}")).collect()
}

#[cfg(test)]
#[path = "capture_tests.rs"]
mod capture_tests;
